starboard/shared/libiamf/iamf_config_reader.cc

```diff
@@ -0,0 +1,287 @@
+constexpr int kFourccOpus = 0x4f707573;
+constexpr int kFourccMp4a = 0x6d703461;
+constexpr int kFourccFlac = 0x664c6143;
+constexpr int kFourccIpcm = 0x6970636d;
+
+constexpr int kLayoutTypeReserved = 1;
+constexpr int kLayoutTypeSpeaker = 2;
+constexpr int kLayoutTypeBinaural = 3;
+
+inline uint32_t ByteSwap(uint32_t x) {
+#if defined(COMPILER_MSVC)
+  return _byteswap_ulong(x);
+#else
+  return __builtin_bswap32(x);
+#endif
+}
+
+// Decodes an Leb128 value and stores it in |value|. Returns the number of bytes
+// read. Returns -1 on error.
+int ReadLeb128Value(const uint8_t* buf, uint32_t* value) {
+  SB_DCHECK(buf);
+  SB_DCHECK(value);
+  *value = 0;
+  bool error = true;
+  size_t i = 0;
+  for (; i < sizeof(uint32_t); ++i) {
+    uint8_t byte = buf[i];
+    *value |= ((byte & 0x7f) << (i * 7));
+    if (!(byte & 0x80)) {
+      error = false;
+      break;
+    }
+  }
+
+  if (error) {
+    return -1;
+  }
+  return i + 1;
+}
+
+int ReadString(const uint8_t* buf, std::string& value) {
+  SB_DCHECK(buf);
+  value = "";
+  value.reserve(128);
+
+  int bytes_read = 0;
+  bool error = true;
+  while (bytes_read < 128 && buf[bytes_read] != '\0') {
+    value.push_back(static_cast<char>(buf[bytes_read]));
+    bytes_read++;
+    error = false;
+  }
+
+  if (error) {
+    return -1;
+  }
+
+  return bytes_read;
+}
+}  // namespace
+
+bool IamfConfigReader::ResetAndRead(scoped_refptr<InputBuffer> input_buffer) {
+  Reset();
+  return Read(input_buffer);
+}
+
+void IamfConfigReader::Reset() {
+  buffer_head_ = 0;
+  mix_presentation_id_ = std::optional<uint32_t>();
+  config_size_ = 0;
+  data_size_ = 0;
+  sample_rate_ = 0;
+  samples_per_buffer_ = 0;
+  sample_size_ = 0;
+}
+
+bool IamfConfigReader::Read(scoped_refptr<InputBuffer> input_buffer) {
+  const uint8_t* buf = input_buffer->data();
+  SB_DCHECK(buf);
+
+  bool completed_parsing = false;
+  while (!completed_parsing && buffer_head_ < input_buffer->size()) {
+    if (!ReadOBU(buf, completed_parsing)) {
+      SB_DLOG(INFO) << "Error parsing config OBUs";
+      return false;
+    }
+  }
+
+  SB_CHECK(completed_parsing);
+
+  data_size_ = input_buffer->size() - config_size_;
+  config_obus_.assign(buf, buf + config_size_);
+  data_.assign(buf + config_size_, buf + input_buffer->size());
+
+  return true;
+}
+
+bool IamfConfigReader::ReadOBU(const uint8_t* buf, bool& completed_parsing) {
+  uint8_t obu_type = 0;
+  uint32_t obu_size = 0;
+  uint32_t header_size = 0;
+  if (!ReadOBUHeader(buf, &obu_type, &obu_size, &header_size)) {
+    SB_DLOG(ERROR) << "Error reading OBU header";
+    return false;
+  }
+
+  int next_obu_pos = buffer_head_ + obu_size;
+  int bytes_read = 0;
+
+  uint32_t count_label = 0;
+  std::string str;
+  uint32_t num_sub_mixes = 0;
+
+  switch (static_cast<int>(obu_type)) {
+    case kObuTypeCodecConfig: {
+      sample_rate_ = 0;
+      uint32_t codec_config_id;
+      bytes_read = ReadLeb128Value(&buf[buffer_head_], &codec_config_id);
```

Comment:
Just want to double check if there is logic here to ensure that there are enough bytes for `ReadLeb128Value()` to read, instead of reading out of bounds, as we don't pass size remaining to `ReadLeb128Value()`.